File: Unified Portfolio/Unified Portfolio/Python/coinbase_module.py
import os

from dotenv import load_dotenv
from coinbase.rest import RESTClient

class Coin:
    def __init__(self, asset):
        self._coin_value = asset["total_balance_crypto"]
        self._curr_value = asset["total_balance_fiat"]
        self._cost_average = asset["average_entry_price"]["value"]
        self._cost_basis = asset["cost_basis"]["value"]

        diff = self._curr_value - float(self.cost_basis)

        if (diff < 0):
            self._unrealized_return = [abs(diff), False]
        else:
            self._unrealized_return = [diff, True]

    @property
    def coin_value(self):
        return self._coin_value

    @property
    def curr_value(self):
        return self._curr_value
    
    @property
    def cost_average(self):
        return self._cost_average

    @property
    def cost_basis(self):
        return self._cost_basis
    
    @property
    def unrealized_return(self):
        return self._unrealized_return

class CoinbaseBalance:
    def __init__(self):
        self._total_balance = [0, "USD"]
        self._total_cost_basis = [0, "USD"]
        self._assets = {}
        self._cash = 0

        load_dotenv()

        api_key = os.getenv("PUBLIC")
        api_secret = os.getenv("PRIVATE")

        self._client = RESTClient(api_key=api_key, api_secret=api_secret)
        self._uuid = self._client.get_portfolios()["portfolios"][0]["uuid"]

        self.get_portfolio_data()

    def get_portfolio_data(self):
        breakdown = self._client.get_portfolio_breakdown(self._uuid)["breakdown"].to_dict()

        total_balance = breakdown["portfolio_balances"]["total_balance"]

        self._total_balance[0], self._total_balance[1] = total_balance["value"], total_balance["currency"]

        for asset in breakdown["spot_positions"]:
            if (not asset["is_cash"]):
                self._assets[asset["asset"]] = Coin(asset)
            else:
                self._cash = asset["total_balance_fiat"]
            self._total_cost_basis[0] += float(asset["cost_basis"]["value"])

    @property
    def total_balance(self):
        return self._total_balance
    
    @property
    def total_cost_basis(self):
        return self._total_cost_basis

    @property
    def assets(self):
        return self._assets
    
    @property
    def cash(self):
        return self._cash

    # Method for testing through terminal
    def display_data(self):
        print("Assets held in Coinbase:")

        for key, val in self._assets.items():
            print(val.coin_value, val.curr_value, val.cost_average, val.cost_basis, key)

        print(self.cash)

        print("Total portfolio value:", self._total_balance[0], self._total_balance[1])

# main() function for testing
if __name__ == '__main__':
    coin = CoinbaseBalance()
    coin.display_data()
    

File: Unified Portfolio/Unified Portfolio/CoinbaseModule.swift
//
//  CoinbaseModule.swift
//  Unified Portfolio
//
//  Created by Jordan Pauzie on 7/24/25.
//

import Foundation
import KeychainSwift
import SwiftJWT

class CoinbaseBalance: ObservableObject {
    @Published var totalBalance: (Double, String) = (0, "USD")
    @Published var totalCostBasis: (Double, String) = (0, "USD")
    @Published var assets: [String: Coin] = [:]
    @Published var cash: Double = 0.0

    private var apiKey: String?
    private var apiSecret: String?
    private var uuid: String?
    private var jwt: String?

    private let keychain = KeychainSwift()

    init() {
        loadSecrets()
        Task {  
            await getPortfolioData() 
        }
    }

    private func loadSecrets() {
        apiKey = keychain.get("COINBASE_PUBLIC")
        apiSecret = keychain.get("COINBASE_PRIVATE")
    }

    private func getUUID() async -> String? {
        guard let jwt else {   
            return nil 
        }

        let path = "/api/v3/brokerage/portfolios"

        var request = URLRequest(url: URL(string: "https://api.coinbase.com\(path)")!)

        request.httpMethod = "GET"
        request.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            if let httpResponse = response as? HTTPURLResponse {
                print("Status: \(httpResponse.statusCode)")
            }

            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                let portfolios = json["portfolios"] as? [[String: Any]],
                let uuid = portfolios.first?["uuid"] as? String {
                return uuid
            } else { 
                print("Failed to parse JSON for UUID") 
            }
        } catch { 
            print("Request failed: \(error)") 
        }

        return nil
    }

    func getPortfolioData() async {
        loadSecrets()

        guard let apiKey, let apiSecret else {
            print("Missing API credentials")
            return
        }

        var path = "/api/v3/brokerage/portfolios"

        guard let jwtUUID = try? generateCoinbaseJWT(apiKey: apiKey, apiSecret: apiSecret, method: "GET", path: path) else {
            print("Failed to create JWT for portfolio list")
            return
        }

        self.jwt = jwtUUID

        guard let uuid = await getUUID() else {
            print("Could not retrieve UUID")
            return
        }

        path = "/api/v3/brokerage/portfolios/\(uuid)"

        guard let jwtPortfolio = try? generateCoinbaseJWT(apiKey: apiKey, apiSecret: apiSecret, method: "GET", path: path) else {
            print("Failed to create JWT for holdings")
            return
        }

        var request = URLRequest(url: URL(string: "https://api.coinbase.com\(path)")!)

        request.httpMethod = "GET"
        request.setValue("Bearer \(jwtPortfolio)", forHTTPHeaderField: "Authorization")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            if let httpResponse = response as? HTTPURLResponse {
                print("Holdings request status: \(httpResponse.statusCode)")
            }

            if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                guard let breakdown = json["breakdown"] as? [String: Any],
                    let spotPositions = breakdown["spot_positions"] as? [[String: Any]],
                    let portfolioBalance = breakdown["portfolio_balances"] as? [String: Any],
                    let totalPortfolioBalance = portfolioBalance["total_balance"] as? [String: Any],
                    let totalCurrency = totalPortfolioBalance["currency"] as? String
                else {
                    print("Unexpected JSON structure in holdings")
                    return
                }
                
                let totalValue = doubleFrom(totalPortfolioBalance, key: "value")

                var assetDict: [String: Coin] = [:]
                var totalCostBasis: Double = 0.0
                var totalCash: Double = 0.0

                for asset in spotPositions {
                    if let isCash = asset["is_cash"] as? Bool, !isCash {
                        let coin = Coin(from: asset)
                        let name = asset["asset"] as? String
                        assetDict[name!] = coin
                    } else if asset["total_balance_fiat"] is Double {
                        totalCash = doubleFrom(asset, key: "total_balance_fiat")
                    }

                    if let costBasisDict = asset["cost_basis"] as? [String: Any] {
                        totalCostBasis += doubleFrom(costBasisDict, key: "value")
                    }
                }

                DispatchQueue.main.async {
                    self.totalBalance = (totalValue, totalCurrency)
                    self.totalCostBasis = (totalCostBasis, totalCurrency)
                    self.assets = assetDict
                    self.cash = totalCash
                }

                // Testing
                print(self.totalBalance)
                print(self.totalCostBasis)
                print(self.assets)
                print(self.cash)
            } else {
                print("Failed to decode holdings JSON")
            }
        } catch {
            print("Holdings request failed: \(error)")
        }
    }
}
